extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantConfigBean.java

```diff
@@ -1,46 +1,85 @@
 package io.quarkus.oidc.runtime;
 
+import static java.util.Collections.unmodifiableMap;
+
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
 
 import jakarta.enterprise.context.spi.CreationalContext;
 
+import org.jboss.logging.Logger;
+
 import io.quarkus.arc.BeanDestroyer;
 import io.quarkus.oidc.OidcTenantConfig;
+import io.quarkus.runtime.configuration.ConfigurationException;
 import io.smallrye.mutiny.Uni;
 
 public class TenantConfigBean {
 
+    private static final Logger LOG = Logger.getLogger(TenantConfigBean.class);
+
     private final Map<String, TenantConfigContext> staticTenantsConfig;
     private final Map<String, TenantConfigContext> dynamicTenantsConfig;
     private final TenantConfigContext defaultTenant;
-    private final Function<OidcTenantConfig, Uni<TenantConfigContext>> tenantConfigContextFactory;
+    private final TenantContextFactory tenantContextFactory;
+
+    @FunctionalInterface
+    public interface TenantContextFactory {
+        Uni<TenantConfigContext> create(OidcTenantConfig oidcTenantConfig, boolean dynamicTenant, String tenantId);
+    }
 
     public TenantConfigBean(
             Map<String, TenantConfigContext> staticTenantsConfig,
-            Map<String, TenantConfigContext> dynamicTenantsConfig,
             TenantConfigContext defaultTenant,
-            Function<OidcTenantConfig, Uni<TenantConfigContext>> tenantConfigContextFactory) {
-        this.staticTenantsConfig = staticTenantsConfig;
-        this.dynamicTenantsConfig = dynamicTenantsConfig;
+            TenantContextFactory tenantContextFactory) {
+        this.staticTenantsConfig = new ConcurrentHashMap<>(staticTenantsConfig);
+        this.dynamicTenantsConfig = new ConcurrentHashMap<>();
         this.defaultTenant = defaultTenant;
-        this.tenantConfigContextFactory = tenantConfigContextFactory;
+        this.tenantContextFactory = tenantContextFactory;
     }
 
-    public Map<String, TenantConfigContext> getStaticTenantsConfig() {
-        return staticTenantsConfig;
+    public Uni<TenantConfigContext> createTenantContext(OidcTenantConfig oidcConfig, boolean dynamicTenant) {
+        var tenantId = oidcConfig.getTenantId().orElseThrow();
+        if (dynamicTenant && oidcConfig.logout.backchannel.path.isPresent()) {
+            throw new ConfigurationException(
+                    "BackChannel Logout is currently not supported for dynamic tenants (tenant ID: " + tenantId + ")");
+        }
+        var tenants = dynamicTenant ? dynamicTenantsConfig : staticTenantsConfig;
+        var tenant = tenants.get(tenantId);
+        if (tenant == null || !tenant.ready) {
+            LOG.tracef("Creating %s tenant config for %s", dynamicTenant ? "dynamic" : "static", tenantId);
+            Uni<TenantConfigContext> uniContext = tenantContextFactory.create(oidcConfig, dynamicTenant, tenantId);
+            return uniContext.onItem().transform(
```

Comment:
It seems that concurrent requests could execute the context-creation simultaneously.

I was thinking, whether we should keep track of the currently being created contexts by tenant-ID (like `ConcurrentMap<String, Uni<TenantConfigContext>>`, containing the `Uni` including the `.transform()`) so that only one creation per tenant ID can happen at any time. I'm not sure though whether it's possible to have different values in `OidcTenantConfig`. Would be something for a separate PR though. WDYT?

Reply:
Definitely for another PR if we really need it, keeping Uni instances might impact memoize behavior as far as I recall from Clement's comments on some other PR. I'd not worry about this optimization for now, thanks

Reply:
Question you are asking is: is is better to create 2+ dynamic tenants or lock/synchronize with another request. I don't think it is an issue that tenant config context is created twice as long as result is idempotent. On the other hand if you synchronized/locked, wouldn't it mean that local caches are going to be invalidated? I don't know how it works, but I am always cautious using synchronization after my proposal here https://github.com/quarkusio/quarkus/pull/37795#issuecomment-1860666461